src/lib.rs

```diff
@@ -1,7 +1,1242 @@
+        }
+    }
+
+    fn master() -> Self {
+        ChildIndex::from_index(0)
+    }
+
+    fn to_index(&self) -> u32 {
+        match self {
+            &ChildIndex::Hardened(i) => i + (1 << 31),
+            &ChildIndex::NonHardened(i) => i,
+        }
+    }
+}
+
+/// A chain code
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct ChainCode([u8; 32]);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct DiversifierIndex(pub [u8; 11]);
+
+impl DiversifierIndex {
+    fn new() -> Self {
+        DiversifierIndex([0; 11])
+    }
+
+    pub fn increment(&mut self) -> Result<(), ()> {
+        let mut k = 0;
+        loop {
+            self.0[k] = self.0[k].wrapping_add(1);
+            if self.0[k] != 0 {
+                // No overflow
+                return Ok(());
+            }
+            // Overflow
+            k += 1;
+            if k == 11 {
+                return Err(());
+            }
+        }
+    }
+}
+
+/// A key used to derive diversifiers for a particular child key
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct DiversifierKey([u8; 32]);
+
+impl DiversifierKey {
+    fn master(sk_m: &[u8]) -> Self {
+        let mut dk_m = [0u8; 32];
+        dk_m.copy_from_slice(&prf_expand(sk_m, &[0x10]).as_bytes()[..32]);
+        DiversifierKey(dk_m)
+    }
+
+    fn derive_child(&self, i_l: &[u8]) -> Self {
+        let mut dk = [0u8; 32];
+        dk.copy_from_slice(&prf_expand_vec(i_l, &[&[0x16], &self.0]).as_bytes()[..32]);
+        DiversifierKey(dk)
+    }
+
+    /// Returns the first index starting from j that generates a valid
+    /// diversifier, along with the corresponding diversifier. Returns
+    /// an error if the diversifier space is exhausted.
+    fn diversifier(&self, mut j: DiversifierIndex) -> Result<(DiversifierIndex, Diversifier), ()> {
+        let ff = FF1::<Aes256>::new(&self.0, 2).unwrap();
+        loop {
+            // Generate d_j
+            let enc = ff.encrypt(&[], &BinaryNumeralString::from_bytes_le(&j.0[..]))
+                .unwrap();
+            let mut d_j = [0; 11];
+            d_j.copy_from_slice(&enc.to_bytes_le());
+            let d_j = Diversifier(d_j);
+
+            // Return (j, d_j) if valid, else increment j and try again
+            match d_j.g_d::<Bls12>(&JUBJUB) {
+                Some(_) => return Ok((j, d_j)),
+                None => if j.increment().is_err() {
+                    return Err(());
+                },
+            }
+        }
+    }
+}
+
+/// A Sapling extended spending key
+#[derive(Clone)]
+pub struct ExtendedSpendingKey {
+    depth: u8,
+    parent_fvk_tag: FVKTag,
+    child_index: ChildIndex,
+    chain_code: ChainCode,
+    pub expsk: ExpandedSpendingKey<Bls12>,
+    dk: DiversifierKey,
+}
+
+// A Sapling extended full viewing key
+pub struct ExtendedFullViewingKey {
+    depth: u8,
+    parent_fvk_tag: FVKTag,
+    child_index: ChildIndex,
+    chain_code: ChainCode,
+    pub fvk: FullViewingKey<Bls12>,
+    dk: DiversifierKey,
+}
```

Comment:
ExtendedSpendingKey and ExtendedFullViewingKey share all of their fields except for ```pub expsk: ExpandedSpendingKey<Bls12>,``` and ```pub fvk: FullViewingKey<Bls12>,``` it would be nice if we could pull out the shared fields into another struct and then to encapsulate some of the shared logic. e.g.,
```
#[derive(Clone, PartialEq)]
struct KeyInfo {
    depth: u8,
    parent_fvk_tag: FVKTag,
    child_index: ChildIndex,
    chain_code: ChainCode,
    dk: DiversifierKey,
}

impl KeyInfo {
    fn debug_fmt(&self, f: &mut std::fmt::Formatter, type_name: &str) -> Result<(), std::fmt::Error> {
        write!(
            f,
            "{}(d = {}, tag_p = {:?}, i = {:?})",
            type_name, self.depth, self.parent_fvk_tag, self.child_index
        )
    }
    
    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
        let depth = reader.read_u8()?;
        let mut tag = [0; 4];
        reader.read_exact(&mut tag)?;
        let i = reader.read_u32::<LittleEndian>()?;
        let mut c = [0; 32];
        reader.read_exact(&mut c)?;
        let mut dk = [0; 32];
        reader.read_exact(&mut dk)?;

        Ok(KeyInfo {
            depth,
            parent_fvk_tag: FVKTag(tag),
            child_index: ChildIndex::from_index(i),
            chain_code: ChainCode(c),
            dk: DiversifierKey(dk),
        })
    }

    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
        writer.write_u8(self.depth)?;
        writer.write_all(&self.parent_fvk_tag.0)?;
        writer.write_u32::<LittleEndian>(self.child_index.to_index())?;
        writer.write_all(&self.chain_code.0)?;

        writer.write_all(&self.dk.0)?;
        Ok(())
    }
}
```
This would require some corresponding changes below, and as is breaks the order of the serialized/deserialized fields. 
Q: Is the order the fields are serialized part of the spec or do they just need to be consistent?

Reply:
The serialization order is part of the spec.